Create multi region article (#3646)
pulumi/pulumi-hugo: themes/default/content/blog/scaling-apps-across-multiple-regions/index.md

Pairing Pulumi Stacks with ESC (Environments, Secrets, and Configuration) results in a flexible deployment strategy. The combination empowers users to segregate configurations for each environment, manage secrets securely, and maintain a centralized store for configuration data. This ensures a clean separation of concerns, contributing to improved security and maintainability in multi-environment deployments. ESC is a pivotal aspect of managing stacks. The idea is to store configuration settings separately from your code, ensuring that each stack has configuration values tailored to its specific needs.

To create and manage stacks for different regions, you can go through the following steps:

### Define Regions and Stacks

> A strategy if you want to deploy to three regions: `production westus2`, `production westeurope`, and `staging southeastasia`, is to create three corresponding stacks, namely: `Pulumi.prodwu2.yaml`, `Pulumi.prodwe.yaml`, and `Pulumi.stagingsea.yaml`. Each stack will contain configurations specific to its region.

Reviewer: Let's rework this a bit to be something more along the lines of "Let's look at an example of deploying to three different regions..." We want to let readers know that in this particular example, you're using stacks to accomplish TWO different things at the same time: 1) deploying to different environments, and 2) deploying to different regions

import * as azure from "@pulumi/azure";

// Define the configuration
const config = new pulumi.Config();
const appServiceTier = config.require("appServiceTier");
const appServiceSize = config.require("appServiceSize");
const environment = config.require("environment");
const region = config.require("region");

const planName = `my-app-service-plan-${environment}`;
const resourceGroupName = `my-resource-group-${environment}`;

// Create the resource group
const resourceGroup = new azure.core.ResourceGroup(resourceGroupName, {
    location: region,
});

// Create the App Service Plan
const appServicePlan = new azure.appservice.Plan(planName, {
    resourceGroupName: resourceGroup.name,
    kind: "Linux",
    reserved: true,
    sku: {
        tier: appServiceTier,
        size: appServiceSize,
    },
});
```

### Deploy Stacks

Deploy each stack to its respective region

```bash
pulumi up --stack prodwu2
pulumi up --stack prodwe
pulumi up --stack stagingsea
```

Now, you have seen how to use stacks to represent regions and environments. With this approach, the dynamism happens through your configuration and your code doesn’t have to change per region.

> To see a full application with more detail, click here.

Reviewer: Needs a link

> ## Implementing High Availability

Reviewer: Let's change this to "Considerations for High Availability" or similar, and reword the following paragraph just a bit. This section is more about the different aspects that architects must consider, and less about actual implementation.

- **Deploying to different cloud provider accounts:** Stacks can also be used to automate deployments across several cloud provider accounts. Even when working with different cloud platforms, your deployment process can be unified and consistent.

Pairing Pulumi Stacks with ESC (Environments, Secrets, and Configuration) results in a flexible deployment strategy. The combination empowers users to segregate configurations for each environment, manage secrets securely, and maintain a centralized store for configuration data. This ensures a clean separation of concerns, contributing to improved security and maintainability in multi-environment deployments. ESC is a pivotal aspect of managing stacks. The idea is to store configuration settings separately from your code, ensuring that each stack has configuration values tailored to its specific needs.

In the upcoming demonstration, we'll illustrate how to deploy applications across diverse environments and regions. To create and manage stacks for different regions, you can go through the following steps:

> ### Define regions and stacks

Reviewer: I'd like you to rethink this section with an emphasis on ESC. Come up with one or more environments that would simplify deploying stacks to multiple regions. Show the stack configurations with and without ESC, showing that ESC is demonstrably better (makes the stack configuration files much simpler). I'm happy to do some brainstorming with you on this, if you'd like.

Let's look at an example of deploying to three distinct regions — `production westus2`, `production westeurope`, and `staging southeastasia` — using stacks. In this scenario, we will use three specific stacks: `Pulumi.prodwu2.yaml`, `Pulumi.prodwe.yaml`, and `Pulumi.stagingsea.yaml`. Each stack serves not only to organize deployments across different environments but also to cater to the unique configurations required for specific regions.

You can create your stacks by running:

> ```bash

Reviewer: I've added suggestions below to fix the `pulumi stack init` commands, which only take a stack name (not the name of the stack configuration file). You might consider using longer or more descriptive stack names here (not a requirement, just a thought).

> pulumi stack init Pulumi.prodwu2.yaml

Reviewer: ```suggestion
pulumi stack init prodwu2
```

> pulumi stack init Pulumi.prodwe.yaml

Reviewer: ```suggestion
pulumi stack init prodwe
```

> pulumi stack init Pulumi.stagungsea.yaml

Reviewer: ```suggestion
pulumi stack init stagungsea
```

meta_image: meta.png
authors:
    - adora-nwodo

# At least one tag is required. Lowercase, hyphen-delimited is recommended.
tags:
    - multi-region-scaling
    - cloud-deployment
    - application-scalability
    - distributed-systems
    - geographic-distribution
    - high-availability

---

As a team building a distributed cloud service that will be used by different people around the world, you will need strategies to cater to users globally, ensuring uninterrupted service even in the face of disruptions. Have you ever wondered how businesses operate and deliver high-level performance across different regions?  Well, that is the result of scaling applications across multiple regions --- in a bid to ensure constant access and flexibility, organizations distribute their applications and databases across various geographic regions so that if one part faces issues, the service remains available elsewhere.

In this article, we will explore the significance of multi-region scaling and show using Pulumi for multi-region deployment.

## Understanding multi-region scaling

> When distributed systems adapt (or implement) multi-region scaling, they already position themselves for high availability because the redundancy minimizes downtime and ensures that the service stays up and running. If one region experiences issues, such as server downtime or network problems, the application remains accessible to users in other regions. This also leads to better performance in their systems because a fast and responsive experience for users across the globe is achieved.

Reviewer: ```suggestion
When distributed systems adopt (or implement) multi-region scaling, they position themselves for high availability because the redundancy minimizes downtime and ensures that the service stays up and running. If one region experiences issues, such as server downtime or network problems, the application remains accessible to users in other regions. This also leads to better performance in their systems because a fast and responsive experience for users across the globe is achieved.
```

  multi-region:environment: production
  multi-region:location: westus2
```

**Pulumi.prodwe.yaml:**

```yaml
config:
  multi-region:appServiceTier: PremiumV3
  multi-region:appServiceSize: P1V3
  multi-region:environment: production
  multi-region:location: westeurope
```

**Pulumi.stagingsea.yaml:**

```yaml
config:
  multi-region:appServiceTier: Standard
  multi-region:appServiceSize: S1
  multi-region:environment: staging
  multi-region:location: southeastasia
```

#### Leveraging Pulumi ESC

For configuration values that are shared across stacks, Pulumi ESC allows you to  create them once and reference them in multiple stacks. In the referenced codebase, there are other values e.g. `resourceNamePrefix`, `storageKind`, `storageSkuName`, and `tenantId`. Since these values are the same across all our stacks, we can leverage ESC so that we can declare them once and reuse them everywhere.

{{% notes type="info" %}} Pulumi ESC is a feature of Pulumi Cloud and in order to use ESC, you should be using Pulumi Cloud as your backend. {{% /notes %}}

To leverage Pulumi ESC, you first need to initialize a new environment like this:

```bash
esc env init shared-multi-region
```

> {{% notes type="info" %}} You can use the esc CLI, as shown, or you can use the `pulumi env` command. {{% /notes %}}

Reviewer: ```suggestion
{{% notes type="info" %}} You can use the `esc` CLI, as shown, or you can use the `pulumi env` command. {{% /notes %}}
```

## Considerations for high availability

Once you’ve configured your multi-region deployments, you should consider strategies for high availability to ensure that applications remain operational and accessible even in the face of failures.

### Placing workloads across multiple regions

Effectively distributing workloads across various geographic regions is a key strategy for maintaining high availability. You can employ one of the following methods to distribute traffic across multiple regions:

- **Active/Passive with hot standby:** In this approach, one region actively handles traffic while the other remains on hot standby, meaning the services in the secondary region are always running.

- **Active/Passive with cold standby:** Similar to hot standby, in this case, the services in the secondary region aren't allocated until needed for failover. This minimizes costs but might increase recovery time.

- **Active/Active:** Both regions are active, and requests are load-balanced between them. If one region becomes unavailable, it is taken out of rotation.

> In the context of infrastructure-as-code using Pulumi, these strategies often translate to the management of multiple stacks. In active/hot standby and active/active configurations, there will typically be two Pulumi stacks, each with its allocated set of resources. This allows for the independent management and deployment of resources in each region.

Reviewer: ```suggestion
In the context of infrastructure as code using Pulumi, these strategies directly translate to the management of multiple stacks. In active/hot standby and active/active configurations, there will typically be two Pulumi stacks, each with its allocated set of resources. This allows for the independent management and deployment of resources in each region.
```

- **Disaster recovery:** Be ready for the worst by copying your important data to different places. Have clear steps on how to bring back your data and programs if everything goes wrong. Spread your tools and services in different areas to reduce problems if one place has a big issue. Make sure your system can easily switch to another area if a big problem happens. Plan for disasters by making step-by-step guides on what to do. Test your disaster plans often to be sure they work well and to find any areas that need fixing.

## Security considerations

> It is one thing to multi-scale, it’s another responsibility to secure the infrastructure. Securing a multi-regional infrastructure introduces challenges. In multi-region setups, data privacy can be tricky due to different regulations, like GDPR in Europe, or The California Consumer Privacy Act (CCPA). Using differential privacy and data residency controls are important when handling data according to each region's rules. Pulumi policy as code also enables the proactive design and enforcement of compliance-ready policies. Ensuring compliance in industries like healthcare and finance adds complexity, requiring a tailored approach based on specific regulations.

Reviewer: ```suggestion
It is one thing to multi-scale, it’s another responsibility to secure the infrastructure. Securing a multi-regional infrastructure introduces challenges. In multi-region setups, data privacy can be tricky due to different regulations, like GDPR in Europe, or The California Consumer Privacy Act (CCPA). Using differential privacy and data residency controls are important when handling data according to each region's rules. [Pulumi policy as code](/crossguard) also enables the proactive design and enforcement of compliance-ready policies. Ensuring compliance in industries like healthcare and finance adds complexity, requiring a tailored approach based on specific regulations.
```

Each of these stacks have their own configuration, but there are also some configuration values that are shared across stacks. To obey the DRY (don't repeat yourself) programming rule, you can pair stacks with ESC (Environments, Secrets, and Configuration). This pairing results in a flexible deployment strategy. The combination empowers users to segregate configurations for each environment, manage secrets securely, and maintain a centralized store for configuration data. This ensures a clean separation of concerns, contributing to improved security and maintainability in multi-environment deployments. ESC is a pivotal aspect of managing stacks. The idea is to store configuration settings separately from your code, ensuring that each stack has configuration values tailored to its specific needs.

In the upcoming demonstration, we'll illustrate how to deploy applications across diverse environments and regions. To create and manage stacks for different regions, you can go through the following steps:

### Define regions and stacks

> Let's look at an example of deploying to three distinct regions — `production westus2`, `production westeurope`, and `staging southeastasia` — using stacks. In this scenario, we will use three specific stacks: `Pulumi.prodwu2.yaml`, `Pulumi.prodwe.yaml`, and `Pulumi.stagingsea.yaml`. Each stack serves not only to organize deployments across different environments but also to cater to the unique configurations required for specific regions.

Reviewer: @AdoraNwodo I didn't see these suggested changes in the latest preview, were they overlooked? We referring to Pulumi stack names, we should just use the stack name and _not_ the file name that is created to store that stack's configuration values. These are different things (but related, obviously).

You can create your stacks by running:

```bash
pulumi stack init prodwu2
pulumi stack init prodwe
pulumi stack init stagingsea
```

After creating your stacks, you can update your Pulumi configuration YAML files with their corresponding values.

> **Pulumi.devwe.yaml:**

Reviewer: @AdoraNwodo This content is still here.

```yaml
config:
  multi-region:appServiceTier: Free
  multi-region:appServiceSize: F1
  multi-region:environment: development
  multi-region:location: westeurope

environment:
  imports:
    - shared-multi-region
```

**Pulumi.prodwu2.yaml:**

```yaml
config:
  multi-region:appServiceTier: PremiumV2
  multi-region:appServiceSize: P1V2
  multi-region:environment: production
  multi-region:location: westus2
```

**Pulumi.prodwe.yaml:**

```yaml
config:
  multi-region:appServiceTier: PremiumV3
  multi-region:appServiceSize: P1V3
  multi-region:environment: production
  multi-region:location: westeurope
```

**Pulumi.stagingsea.yaml:**

```yaml
config:
  multi-region:appServiceTier: Standard
  multi-region:appServiceSize: S1
  multi-region:environment: staging
  multi-region:location: southeastasia
```

> #### Leveraging ESC

Reviewer: @AdoraNwodo Another missed change.

> For configuration values that are shared across stacks, you should use ESC so that you can create them once and reference them everywhere. In the referenced codebase, there are other values e.g. `resourceNamePrefix`, `storageKind`, `storageSkuName`, and `tenantId`. Since these values are the same across all our stacks, we can leverage ESC so that we can declare them once and reuse them everywhere.

Reviewer: @AdoraNwodo Another missed suggested change

{{% notes type="info" %}} Pulumi ESC is a feature of Pulumi Cloud and in order to use ESC, you should be using Pulumi Cloud as your backend. {{% /notes %}}

> To leverage ESC, you first need to initialize a new environment like this:

Reviewer: @AdoraNwodo Same as the others, looks like you didn't expand all the conversation on this page and missed a bunch of suggested edits.

Once you have referenced your ESC environment in your Pulumi config, you can update your Pulumi code to use the configurations from the Pulumi stack:

```typescript
import * as pulumi from "@pulumi/pulumi";
import * as azure from "@pulumi/azure";

// Define the configuration
const stackName: string = pulumi.getStack();
const config = new pulumi.Config();

const appServiceTier = config.require("appServiceTier");
const appServiceSize = config.require("appServiceSize");
const resourceNamePrefix = config.require("resourceNamePrefix"); // gotten from ESC
const storageKind = config.require("storageKind"); // gotten from ESC

// ... more code

// Create an Azure Resource Group
const resourceGroup = new azure.resources.ResourceGroup(resourceGroupName, {
    resourceGroupName: resourceGroupName,
    location: location,
    tags
});

// Create an Azure Storage Account
const storageAccount = new azure.storage.StorageAccount(storageAccountName, {
    accountName: storageAccountName,
    resourceGroupName: resourceGroup.name,
    kind: storageKind,
    location: location,
    sku: {
        name: storageSkuName
    },
    tags
});

// ... more code
```

### Deploy stacks

> After writing your infrastructure code, deploy each stack to its respective region

Reviewer: ```suggestion
Your Pulumi stacks should be ready to deploy now---you've updated the code to use configuration values that either come from the stack or from Pulumi ESC, and you've done both the per-stack configurations and the Pulumi ESC configuration values. Now it's just a matter of running `pulumi up`:
```

To see a full application with more detail, click [here](https://github.com/pulumi/pulumitv/tree/master/multi-region).

## Considerations for high availability

Once you’ve configured your multi-region deployments, you should consider strategies for high availability to ensure that applications remain operational and accessible even in the face of failures.

### Placing workloads across multiple regions

Effectively distributing workloads across various geographic regions is a key strategy for maintaining high availability. You can employ one of the following methods to distribute traffic across multiple regions:

- **Active/Passive with hot standby:** In this approach, one region actively handles traffic while the other remains on hot standby, meaning the services in the secondary region are always running.

- **Active/Passive with cold standby:** Similar to hot standby, in this case, the services in the secondary region aren't allocated until needed for failover. This minimizes costs but might increase recovery time.

- **Active/Active:** Both regions are active, and requests are load-balanced between them. If one region becomes unavailable, it is taken out of rotation.

> 

Reviewer: Add a paragraph here explaining that in active/hot standby and active/active, there will be two Pulumi stacks with allocated resources. In active/cold standby, there might be a stack for the cold standby site, but it won't have any resources allocated (part of the failover process will be running a `pulumi up` on the cold standby region).
